daft/catalog/__gravitino.py

```diff
@@ -76,37 +76,52 @@ def _get_table(self, ident: Identifier) -> GravitinoTable:
     ###
 
     def _list_namespaces(self, pattern: str | None = None) -> list[Identifier]:
-        raise NotImplementedError("Gravitino list_namespaces not yet supported.")
+        if pattern is None or pattern == "":
+            return [
+                Identifier.from_str(namespace)
+                for cat in self._inner.list_catalogs()
+                for namespace in self._inner.list_namespaces(cat)
+            ]
+        # If pattern is provided, treat it as a catalog name
+        catalog_name = pattern
+        return [Identifier.from_str(namespace) for namespace in self._inner.list_namespaces(catalog_name)]
 
     def _list_tables(self, pattern: str | None = None) -> list[Identifier]:
         if pattern is None or pattern == "":
             return [
                 Identifier.from_str(tbl)
                 for cat in self._inner.list_catalogs()
-                for schema in self._inner.list_schemas(cat)
-                for tbl in self._inner.list_tables(schema)
+                for namespace in self._inner.list_namespaces(cat)
+                for tbl in self._inner.list_tables(namespace)
             ]
         num_namespaces = pattern.count(".")
         if num_namespaces == 0:
             catalog_name = pattern
             return [
                 Identifier.from_str(tbl)
-                for schema in self._inner.list_schemas(catalog_name)
-                for tbl in self._inner.list_tables(schema)
+                for namespace in self._inner.list_namespaces(catalog_name)
+                for tbl in self._inner.list_tables(namespace)
             ]
         elif num_namespaces == 1:
-            schema_name = pattern
-            return [Identifier.from_str(tbl) for tbl in self._inner.list_tables(schema_name)]
+            namespace_name = pattern
+            return [Identifier.from_str(tbl) for tbl in self._inner.list_tables(namespace_name)]
         else:
             raise ValueError(
-                f"Unrecognized catalog name or schema name, expected a '.'-separated namespace but received: {pattern}"
+                f"Unrecognized catalog name or namespace name, expected a '.'-separated namespace but received: {pattern}"
             )
 
     ###
     # has_.*
     ###
     def _has_namespace(self, ident: Identifier) -> bool:
-        raise NotImplementedError("Gravitino has_namespace not yet supported.")
+        ident_str = str(ident)
+        # Namespace should be in format "catalog.schema"
+        if ident_str.count(".") != 1:
+            return False
+        catalog_name = ident_str.split(".")[0]
+        # List namespaces for the catalog and check if our identifier is in the list
+        namespaces = self._inner.list_namespaces(catalog_name)
+        return ident_str in namespaces
```

Comment:
Ident is an iterable of str, so you don't need to split or check. You can do things like `len(ident) != 2` as well as just get the parts with indexing `ident[0]`.